File: ai.devictoria.shop/services/authservice/app/services/naver_auth_service.py
"""
네이버 인증 서비스
"""
import httpx
from typing import Optional
from app.config import NaverProperties
from app.dto.provider import NaverTokenResponse, NaverUserInfo


class NaverAuthService:
    """네이버 OAuth 인증 서비스"""
    
    def __init__(self, naver_properties: NaverProperties):
        self.naver_properties = naver_properties
    
    def get_naver_token(self, authorization_code: str, state: Optional[str] = None) -> NaverTokenResponse:
        """네이버 인증 코드로 토큰 요청"""
        print("[NaverAuthService] 네이버 토큰 요청 시작")
        
        form_data = {
            "grant_type": "authorization_code",
            "client_id": self.naver_properties.client_id,
            "client_secret": self.naver_properties.client_secret,
            "redirect_uri": self.naver_properties.redirect_uri,
            "code": authorization_code
        }
        
        if state:
            form_data["state"] = state
        
        try:
            print("[NaverAuthService] 네이버 API에 토큰 요청 전송 중...")
            with httpx.Client() as client:
                response = client.post(
                    self.naver_properties.token_uri,
                    data=form_data,
                    headers={"Content-Type": "application/x-www-form-urlencoded"}
                )
                response.raise_for_status()
                token_data = response.json()
            
            token_response = NaverTokenResponse(**token_data)
            print("[NaverAuthService] 네이버 토큰 요청 성공!")
            return token_response
            
        except Exception as e:
            print(f"[NaverAuthService] 네이버 토큰 요청 실패: {e}")
            raise RuntimeError(f"네이버 토큰 요청 실패: {e}")
    
    def get_naver_user_info(self, access_token: str) -> NaverUserInfo:
        """네이버 액세스 토큰으로 사용자 정보 조회"""
        print("[NaverAuthService] 네이버 사용자 정보 조회 시작")
        
        try:
            print("[NaverAuthService] 네이버 API에 사용자 정보 요청 전송 중...")
            with httpx.Client() as client:
                response = client.get(
                    self.naver_properties.user_info_uri,
                    headers={"Authorization": f"Bearer {access_token}"}
                )
                response.raise_for_status()
                user_data = response.json()
            
            user_info = NaverUserInfo(**user_data)
            print("[NaverAuthService] 네이버 사용자 정보 조회 성공!")
            return user_info
            
        except Exception as e:
            print(f"[NaverAuthService] 네이버 사용자 정보 조회 실패: {e}")
            raise RuntimeError(f"네이버 사용자 정보 조회 실패: {e}")
    
    def login_with_authorization_code(self, authorization_code: str, state: Optional[str] = None) -> NaverUserInfo:
        """인증 코드로 네이버 로그인 처리"""
        print("[NaverAuthService] 인증 코드로 로그인 처리 시작")
        token_response = self.get_naver_token(authorization_code, state)
        return self.get_naver_user_info(token_response.access_token)
    
    def login_with_access_token(self, access_token: str) -> NaverUserInfo:
        """액세스 토큰으로 네이버 로그인 처리"""
        print("[NaverAuthService] 액세스 토큰으로 로그인 처리 시작")
        return self.get_naver_user_info(access_token)

File: api.devictoria.shop/src/main/java/shop/devictoria/api/oauthservice/service/auth/provider/NaverAuthService.java
package shop.devictoria.api.service.auth.provider;

import com.fasterxml.jackson.core.JsonProcessingException;
import com.fasterxml.jackson.databind.ObjectMapper;
import lombok.RequiredArgsConstructor;
import org.springframework.data.redis.core.RedisTemplate;
import org.springframework.http.HttpHeaders;
import org.springframework.http.MediaType;
import org.springframework.stereotype.Service;
import org.springframework.util.LinkedMultiValueMap;
import org.springframework.util.MultiValueMap;
import org.springframework.web.reactive.function.client.WebClient;
import shop.devictoria.api.config.NaverProperties;
import shop.devictoria.api.dto.auth.provider.NaverTokenResponse;
import shop.devictoria.api.dto.auth.provider.NaverUserInfo;

import java.util.concurrent.TimeUnit;

@Service
@RequiredArgsConstructor
public class NaverAuthService {
    
    private final NaverProperties naverProperties;
    private final WebClient.Builder webClientBuilder;
    private final RedisTemplate<String, String> redisTemplate;
    private final ObjectMapper objectMapper = new ObjectMapper();
    
    /**
     * 네이버 인증 코드로 토큰 요청
     */
    public NaverTokenResponse getNaverToken(String authorizationCode, String state) {
        System.out.println("[NaverAuthService] 네이버 토큰 요청 시작");
        System.out.println("[NaverAuthService] 인증 코드: " + authorizationCode);
        System.out.println("[NaverAuthService] State: " + (state != null ? state : "null"));
        System.out.println("[NaverAuthService] Client ID: " + naverProperties.getClientId());
        System.out.println("[NaverAuthService] Redirect URI: " + naverProperties.getRedirectUri());
        System.out.println("[NaverAuthService] Token URI: " + naverProperties.getTokenUri());
        
        MultiValueMap<String, String> formData = new LinkedMultiValueMap<>();
        formData.add("grant_type", "authorization_code");
        formData.add("client_id", naverProperties.getClientId());
        formData.add("client_secret", naverProperties.getClientSecret());
        formData.add("redirect_uri", naverProperties.getRedirectUri());
        formData.add("code", authorizationCode);
        // 네이버는 state 파라미터가 필수이므로, 없으면 빈 문자열 사용
        formData.add("state", state != null ? state : "");
        
        try {
            WebClient webClient = webClientBuilder.build();
            
            System.out.println("[NaverAuthService] 네이버 API에 토큰 요청 전송 중...");
            NaverTokenResponse response = webClient.post()
                    .uri(naverProperties.getTokenUri())
                    .contentType(MediaType.APPLICATION_FORM_URLENCODED)
                    .bodyValue(formData)
                    .retrieve()
                    .bodyToMono(NaverTokenResponse.class)
                    .block();
            
            if (response.getError() != null) {
                System.out.println("[NaverAuthService] 네이버 토큰 요청 실패: " + response.getError() + " - " + response.getErrorDescription());
                throw new RuntimeException("네이버 토큰 요청 실패: " + response.getErrorDescription());
            }
            
            System.out.println("[NaverAuthService] 네이버 토큰 요청 성공!");
            System.out.println("[NaverAuthService] 받은 Access Token: " + (response.getAccessToken() != null ? response.getAccessToken().substring(0, Math.min(20, response.getAccessToken().length())) + "..." : "null"));
            return response;
            
        } catch (Exception e) {
            System.out.println("[NaverAuthService] 네이버 토큰 요청 실패: " + e.getMessage());
            e.printStackTrace();
            throw new RuntimeException("네이버 토큰 요청 실패", e);
        }
    }
    
    /**
     * 네이버 액세스 토큰으로 사용자 정보 조회
     */
    public NaverUserInfo getNaverUserInfo(String accessToken) {
        System.out.println("[NaverAuthService] 네이버 사용자 정보 조회 시작");
        System.out.println("[NaverAuthService] 사용할 Access Token: " + (accessToken != null ? accessToken.substring(0, Math.min(20, accessToken.length())) + "..." : "null"));
        System.out.println("[NaverAuthService] User Info URI: " + naverProperties.getUserInfoUri());
        
        try {
            WebClient webClient = webClientBuilder.build();
            
            System.out.println("[NaverAuthService] 네이버 API에 사용자 정보 요청 전송 중...");
            NaverUserInfo userInfo = webClient.get()
                    .uri(naverProperties.getUserInfoUri())
                    .header(HttpHeaders.AUTHORIZATION, "Bearer " + accessToken)
                    .retrieve()
                    .bodyToMono(NaverUserInfo.class)
                    .block();
            
            if (!"00".equals(userInfo.getResultCode())) {
                System.out.println("[NaverAuthService] 네이버 사용자 정보 조회 실패: " + userInfo.getMessage());
                throw new RuntimeException("네이버 사용자 정보 조회 실패: " + userInfo.getMessage());
            }
            
            System.out.println("[NaverAuthService] 네이버 사용자 정보 조회 성공!");
            if (userInfo.getResponse() != null) {
                System.out.println("[NaverAuthService] 사용자 ID: " + userInfo.getResponse().getId());
                System.out.println("[NaverAuthService] 이메일: " + userInfo.getResponse().getEmail());
                System.out.println("[NaverAuthService] 이름: " + userInfo.getResponse().getName());
                System.out.println("[NaverAuthService] 닉네임: " + userInfo.getResponse().getNickname());
            }
            return userInfo;
            
        } catch (Exception e) {
            System.out.println("[NaverAuthService] 네이버 사용자 정보 조회 실패: " + e.getMessage());
            e.printStackTrace();
            throw new RuntimeException("네이버 사용자 정보 조회 실패", e);
        }
    }
    
    /**
     * 인증 코드로 네이버 로그인 처리 (토큰 요청 + 사용자 정보 조회)
     */
    public NaverUserInfo loginWithAuthorizationCode(String authorizationCode, String state) {
        System.out.println("[NaverAuthService] 인증 코드로 로그인 처리 시작");
        // 1. 인증 코드로 토큰 요청
        NaverTokenResponse tokenResponse = getNaverToken(authorizationCode, state);
        
        // 2. 토큰으로 사용자 정보 조회
        NaverUserInfo userInfo = getNaverUserInfo(tokenResponse.getAccessToken());
        
        // 3. 네이버 토큰을 Redis에 저장
        if (userInfo.getResponse() != null) {
            saveNaverTokenToRedis(userInfo.getResponse().getId(), tokenResponse);
        }
        
        return userInfo;
    }
    
    /**
     * 네이버 토큰을 Redis에 저장
     */
    private void saveNaverTokenToRedis(String userId, NaverTokenResponse tokenResponse) {
        try {
            String key = "naver:token:" + userId;
            String tokenJson = objectMapper.writeValueAsString(tokenResponse);
            
            // 네이버 액세스 토큰 만료 시간 설정 (초 단위)
            long expiration = tokenResponse.getExpiresIn() != null ? 
                tokenResponse.getExpiresIn() : 3600; // 기본 1시간
            
            redisTemplate.opsForValue().set(key, tokenJson, expiration, TimeUnit.SECONDS);
            
            System.out.println("========================================");
            System.out.println("[NaverAuthService] ✅ 네이버 토큰을 Redis에 저장했습니다");
            System.out.println("[NaverAuthService] Redis Key: " + key);
            System.out.println("[NaverAuthService] 만료 시간: " + expiration + "초");
            System.out.println("[NaverAuthService] 저장된 네이버 Access Token: " + 
                (tokenResponse.getAccessToken() != null ? tokenResponse.getAccessToken().substring(0, Math.min(20, tokenResponse.getAccessToken().length())) + "..." : "null"));
            System.out.println("[NaverAuthService] 저장된 네이버 Refresh Token: " + 
                (tokenResponse.getRefreshToken() != null ? tokenResponse.getRefreshToken().substring(0, Math.min(20, tokenResponse.getRefreshToken().length())) + "..." : "null"));
            System.out.println("(이 토큰은 Redis에만 저장되고, 프론트엔드로는 반환되지 않습니다)");
            System.out.println("========================================");
        } catch (JsonProcessingException e) {
            System.out.println("[NaverAuthService] Redis 저장 실패: " + e.getMessage());
            e.printStackTrace();
        } catch (Exception e) {
            System.out.println("[NaverAuthService] Redis 저장 중 오류 발생: " + e.getMessage());
            e.printStackTrace();
        }
    }
    
    /**
     * 액세스 토큰으로 네이버 로그인 처리 (사용자 정보 조회만)
     */
    public NaverUserInfo loginWithAccessToken(String accessToken) {
        System.out.println("[NaverAuthService] 액세스 토큰으로 로그인 처리 시작");
        return getNaverUserInfo(accessToken);
    }
}

